File: app/src/main/java/com/stemsc/vmkt/Send.java
package com.stemsc.routelist;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.preference.PreferenceManager;
import android.util.Log;

//import com.stemsc.routelist.db.DBHelper;
import com.stemsc.routelist.db.DBHelperDate;
import com.stemsc.routelist.db.DBHelperLog;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.UnknownHostException;

/**
 * Created by ONB on 11.07.2017.
 * Send
 */

class Send extends AsyncTask<String, Void, JSONObject> {

    private static final String TAG = ">>Send>>";
    private String method;
    private String body;
    private int respCode = 999;
    private SharedPreferences sPref;
    private DBHelperLog dbHelperLog;
//    private DBHelper dbHelper;
    private DBHelperDate dbHelperDate;

    Send(Context context){
        sPref = PreferenceManager.getDefaultSharedPreferences(context);
        dbHelperLog = new DBHelperLog(context);
        dbHelperDate = new DBHelperDate(context);
//        dbHelper = new DBHelper(context);
//        Log.d(">>> ", "Send(Context context)");
    }

    @Override
    protected JSONObject doInBackground(String... urls) {
        JSONObject ret = new JSONObject();
        method = urls[0];
        String url = urls[1];
        body = urls[2];
//        Log.d(">>> ", "doInBackground");
        try {
            ret.put("data", loadFromNetwork(url));
            ret.put("code", respCode);
            Log.d(">>> ", method + "=" + url);
            Log.d("<<< ", "(" + ret.getInt("code") + ") " + ret.getString("data"));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return ret;
    }

    private String loadFromNetwork(String urlString) {
        InputStream stream = null;
        String str = "";
        try {
            stream = downloadUrl(urlString);
            if (stream != null)
                str = readIt(stream);
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            if (stream != null) {
                try {
                    stream.close();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
        return str;
    }

    /**
     * @param urlString A string representation of a URL.
     * @return An InputStream retrieved from a successful HttpURLConnection.
     */
    private InputStream downloadUrl(String urlString) throws IOException {

//        if (ServGPS.context != null) context=ServGPS.context;
//        else if (ServTimer.context != null) context=ServTimer.context;
//        else if (MainActivity.context != null) context=MainActivity.context;
//        else if (RoutesActivity.context != null) context=RoutesActivity.context;
//        else if (PointsActivity.context != null) context=PointsActivity.context;
//
//        sPref = PreferenceManager.getDefaultSharedPreferences(context);
//        DBHelper dbHelperLog = new DBHelper(context);

        String MY_TOKEN = sPref.getString("token", "");

        String BASE_URL = (sPref.contains("dev") ? sPref.getString("dev", Utils.WRK) : Utils.WRK);

        URL url = new URL(BASE_URL + "/api/" + urlString);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setConnectTimeout(20000);
        conn.setReadTimeout(20000);
        conn.setRequestMethod(method);
        conn.setDoInput(true);

        assert MY_TOKEN != null;
        if (!MY_TOKEN.equals("")) {
            //  String basicAuth = "Basic " + new String(Base64.encode("driver:1111".getBytes(),Base64.NO_WRAP ));
            String basicAuth = "Token " + MY_TOKEN;
            conn.setRequestProperty("Authorization", basicAuth);
        }
        InputStream ret = null;
        try {
            if (body != null) {
                conn.setDoOutput(true);
                conn.setRequestProperty("Content-Type", "application/json");
                conn.setRequestProperty("Content-Length", "" + Integer.toString(body.getBytes().length));
                OutputStream os = conn.getOutputStream();
                if (os != null) os.write(body.getBytes("UTF-8"));
//            else  ??????????
            }

            conn.connect();
            respCode = conn.getResponseCode();
            switch (respCode) {
//                case 401:
//                    sPref.edit().remove("token").apply();
//                exit(0);
//                    break;
                case 200:
                case 201:
                    ret = conn.getInputStream();
                    dbHelperLog.log("<font color=\"green\">=SEND = "+method+" "+urlString+" ("+respCode+")</font>");
                    if(method.equals("GET")) {
                        dbHelperDate.setTimeUrl(urlString);
                        Log.d(TAG, "dbHelperLog.setTimeUrl = " + urlString);
                    }
                    break;
                default:
                    dbHelperLog.log("<font color=\"red\">=SEND ERROR == "+method+" "+urlString+" ("+respCode+")</font>");
                    Log.d(TAG,"<font color=\"red\">=SEND = respCode="+respCode+"</font>");
//                ret = conn.getInputStream();
            }
        } catch (UnknownHostException e) {
            Log.d(TAG,"===== SEND error = UnknownHostException");
            dbHelperLog.log("<font color=\"red\">=SEND = UnknownHostException="+urlString+"</font>");
            return null;
        } catch (java.net.SocketTimeoutException e) {
            Log.d(TAG,"===== SEND error = SocketTimeoutException");
            dbHelperLog.log("<font color=\"red\">=SEND = SocketTimeoutException="+urlString+"</font>");
            return null;
        } catch (java.io.EOFException e) {
            Log.d(TAG,"===== SEND error = EOFException");
            dbHelperLog.log("<font color=\"red\">=SEND = EOFException="+urlString+"</font>");
            return null;
        } catch (javax.net.ssl.SSLHandshakeException e) {
            Log.d(TAG,"===== SEND error = SSLHandshakeException");
            dbHelperLog.log("<font color=\"red\">=SEND = SSLHandshakeException="+urlString+"</font>");
            return null;
        }
        return ret;
    }

    /**
     * @param stream InputStream containing HTML from targeted site.
     * @return String concatenated according to len parameter.
     */
    private String readIt(InputStream stream) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(stream));
        StringBuilder buffer = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            buffer.append(line);
        }
        return new String(buffer);
    }
}

File: app/src/main/java/com/stemsc/vmkt/Send.kt
package com.stemsc.vmkt

//import android.content.Context
import android.os.AsyncTask
import android.util.Log
//import androidx.preference.PreferenceManager
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.UnknownHostException

const val TAG = "Send"

//internal class Send constructor(context: Context) : AsyncTask<String, Void, JSONObject?>() {
internal class Send : AsyncTask<String, Void, JSONObject?>() {

    //    private var resp: String? = null
//    private val activityReference: WeakReference<MainActivity> = WeakReference(context)

    private var method: String? = null
    private var body: String? = null
    private var respCode = 999
//    private val sPref = PreferenceManager.getDefaultSharedPreferences(context)


    override fun doInBackground(vararg urls: String): JSONObject {
        val ret = JSONObject()
        method = urls[0]
        val url = urls[1]
        body = urls[2]
        try {
            ret.put("data", loadFromNetwork(url))
            ret.put("code", respCode)
            Log.d(TAG, "$method=$url")
            Log.d(TAG, "(" + ret.getInt("code") + ") " + ret.getString("data"))
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        return ret
    }

    private fun loadFromNetwork(urlString: String): String {
        var stream: InputStream? = null
        var str = ""
        try {
            stream = downloadUrl(urlString)
            if (stream != null)
                str = readIt(stream)
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            if (stream != null) {
                try {
                    stream.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }

            }
        }
        return str
    }

    private fun downloadUrl(urlString: String): InputStream? {
/*
        val MY_TOKEN = sPref.getString("token", "")
        val BASE_URL = if (sPref.contains("dev")) sPref.getString("dev", "Utils.WRK") else "Utils.WRK"
*/

        val url = URL("$BASE_URL/api/$urlString")
        val conn = url.openConnection() as HttpURLConnection
        conn.connectTimeout = 20000
        conn.readTimeout = 20000
        conn.requestMethod = method
        conn.doInput = true

//        assert(MY_TOKEN != null)
//        if (MY_TOKEN != "") {
        //  String basicAuth = "Basic " + new String(Base64.encode("driver:1111".getBytes(),Base64.NO_WRAP ));
        val basicAuth = "Token $MY_TOKEN"
        conn.setRequestProperty("Authorization", basicAuth)
//        }
        var ret: InputStream? = null
        try {
            if (body != null) {
                conn.doOutput = true
                conn.setRequestProperty("Content-Type", "application/json")
                conn.setRequestProperty(
                    "Content-Length",
                    "" + body!!.toByteArray().size.toString()
                )
                val os = conn.outputStream
                os?.write(body!!.toByteArray(charset("UTF-8")))
                //            else  ??????????
            }

            conn.connect()
            respCode = conn.responseCode
            when (respCode) {
                //                case 401:
                //                    sPref.edit().remove("token").apply();
                //                exit(0);
                //                    break;
                200, 201 -> {
                    ret = conn.inputStream
                }
                else -> {
//                    dbHelperLog.log("<font color=\"red\">=SEND ERROR == $method $urlString ($respCode)</font>")
                    Log.d(TAG, "===== SEND error = $method - $url - respCode=$respCode")
                }
            }//                ret = conn.getInputStream();
        } catch (e: UnknownHostException) {
            Log.d(TAG, "===== SEND error = UnknownHostException")
            return null
        } catch (e: java.net.SocketTimeoutException) {
            Log.d(TAG, "===== SEND error = SocketTimeoutException")
            return null
        } catch (e: java.io.EOFException) {
            Log.d(TAG, "===== SEND error = EOFException")
            return null
        } catch (e: javax.net.ssl.SSLHandshakeException) {
            Log.d(TAG, "===== SEND error = SSLHandshakeException")
            return null
        }

        return ret
    }

    @Throws(IOException::class)
    private fun readIt(stream: InputStream): String {
        val reader = BufferedReader(InputStreamReader(stream))
        val buffer = StringBuilder()
        var line = reader.readLine()
        while (line  != null) {
            buffer.append(line)
            line = reader.readLine()
        }
        return String(buffer)
    }
/*

    override fun onPreExecute() {
//        val activity = activityReference.get()
//        if (activity == null || activity.isFinishing) return
//        activity.progressBar.visibility = View.VISIBLE
    }



    override fun onPostExecute(result: String?) {

        // execution of result of Long time consuming operation
        val activity = activityReference.get()
        if (activity == null || activity.isFinishing) return
        // access Activity member variables or modify the activity's UI
        activity.progressBar.visibility = View.GONE
        activity.textView.text = result.let { it }
        activity.myVariable = 100
    }

    override fun onProgressUpdate(vararg text: String?) {

        val activity = activityReference.get()
        if (activity == null || activity.isFinishing) return

        Toast.makeText(activity, text.firstOrNull(), Toast.LENGTH_SHORT).show()

    }
*/
}
